fairseq2n/src/fairseq2n/data/multi_data_source.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "fairseq2n/data/multi_data_source.h"
+
+namespace fairseq2n::detail {
+
+multi_data_source::multi_data_source(std::vector<data_pipeline> &&pipelines, index_generator_fn &&index_gen_fn, bool stop_at_shortest)
```

Comment:
Nit: I haven't counted it, but if this line exceeds ~100 chars, please break it.